### PR TITLE
chore: debug test by extending timeout

### PR DESCRIPTION
**Pull request by averikitsch**
Fixes #issue

> It's a good idea to open an issue first for discussion.

- [x] I have followed [Sample Format Guide](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/master/SAMPLE_FORMAT.md)
- [ ] `pom.xml` parent set to latest `shared-configuration`
- [ ] Appropriate changes to README are included in PR
- [ ] API's need to be enabled to test (tell us)
- [ ] Environment Variables need to be set (ask us to set them)
- [ ] **Tests** pass:   `mvn clean verify` **required**
- [ ] **Lint**  passes: `mvn -P lint checkstyle:check` **required**
- [ ] **Static Analysis**:  `mvn -P lint clean compile pmd:cpd-check spotbugs:check` **advisory only**
- [ ] Please **merge** this PR for me once it is approved.

### CODE REVIEW

**Comment by averikitsch** [comment]
@lesv I am trying to fix flaky test #5588. It seems to only be flaky during the nightly tests though. Do you have any suggestions?

**Comment by lesv** [comment]
@averikitsch @kurtisvg has also been looking into these.  He declared victory yesterday. :(

```
#
# unittests
#

[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 299.95 s <<< FAILURE! - in com.google.appengine.samples.DeferredTaskTest
[ERROR] com.google.appengine.samples.DeferredTaskTest.testTaskGetsRun  Time elapsed: 299.947 s  <<< ERROR!
org.junit.runners.model.TestTimedOutException: test timed out after 5 minutes
	at java.lang.Thread.sleep(Native Method)
	at org.quartz.core.QuartzScheduler.shutdown(QuartzScheduler.java:500)
	at org.quartz.impl.StdScheduler.shutdown(StdScheduler.java:202)
	at com.google.appengine.api.taskqueue.dev.LocalTaskQueue.stopScheduler(LocalTaskQueue.java:720)
	at com.google.appengine.api.taskqueue.dev.LocalTaskQueue.stop_(LocalTaskQueue.java:350)
	at com.google.appengine.api.taskqueue.dev.LocalTaskQueue.stop(LocalTaskQueue.java:345)
	at com.google.appengine.tools.development.testing.LocalTaskQueueTestConfig.tearDown(LocalTaskQueueTestConfig.java:220)
	at com.google.appengine.tools.development.testing.LocalServiceTestHelper.tearDownService(LocalServiceTestHelper.java:567)
	at com.google.appengine.tools.development.testing.LocalServiceTestHelper.tearDown(LocalServiceTestHelper.java:534)
	at com.google.appengine.samples.DeferredTaskTest.tearDown(DeferredTaskTest.java:63)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunAfters.invokeMethod(RunAfters.java:46)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)

[ERROR] Errors:
[ERROR]   DeferredTaskTest.tearDown:63 » TestTimedOut test timed out after 5 minutes
[ERROR] Tests run: 15, Failures: 0, Errors: 1, Skipped: 0
```

I have a suspicion.  Let me read some code

**Comment by lesv** [comment]
@averikitsch Kurtis's tool `bltr` runs tests w/ a timeout of 10m or so, so increasing the TO to 10m will definitely fail.

**Comment by kurtisvg** [comment]
Sorry, I declared victory on the "kokoro jobs hitting the timeout" problem. The tests themselves are still broken, they just aren't causing other tests to timeout anymore. 

`btlr` is currently set to a [timeout of 20 minutes](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/4d08f29cf609a16b707efa2ac7358fc0e4d89827/.kokoro/tests/run_tests.sh#L96).

**Comment by lesv** [comment]
@averikitsch You might wish to talk w/ @ludoch when he returns about this.  It's kind of a special snowflake in that until the updated library is released, it should only be run w/ Java 8, not Java 11.  (Not the current problem anyway).

**Review by kurtisvg — APPROVED** [review]

File: unittests/src/test/java/com/google/appengine/samples/BaseTestConfiguration.java [hunk 1]
```diff
@@ -21,5 +21,5 @@
 import org.junit.rules.Timeout;
 
 public abstract class BaseTestConfiguration {
-  @Rule public Timeout testTimeout = new Timeout(5, TimeUnit.MINUTES);
+  @Rule public Timeout testTimeout = new Timeout(10, TimeUnit.MINUTES);
```
Comment (lesv) [inline unittests/src/test/java/com/google/appengine/samples/BaseTestConfiguration.java:24]: Increasing this means that each test can take upto 10m, perhaps we should only do this on the test that is failing?

**Comment by lesv** [comment]
My initial thought is to not update `BaseTestConfiguration.java`, but to update `DeferredTaskTest.java` to no longer inherit from `BaseTestConfiguration`, and add the longer rule there.  I'm still reading, so there will likely be additional comments.

**Comment by lesv** [comment]
In `DeferredTaskTest.java`, the use of `taskRan` isn't threadSafe.  The bug is happening because the original thread can't actually see the change.

**Comment by lesv** [comment]
@averikitsch Let me know if you need a fix, or you remember enough from the first book we read.

**Review by lesv — CHANGES_REQUESTED** [review]
go/ej3e Item#78
Fix goes into src/test/java/com/google/appengine/samples/DeferredTaskTest.java which looks a lot like the first example.

**Comment by lesv** [comment]
I may have spoken too soon, but it's related.

**Comment by lesv** [comment]
What appears to be happening is that in `DeferredTaskTest.testTaskGetsRun()` either `QueueFactory.getDefaultQueue().add(
        TaskOptions.Builder.withPayload(new MyTask()));` or  `assertTrue(latch.await(5, TimeUnit.SECONDS));` never completes.  My guess is that it's a problem between the updated java libraries and GAE testing libraries and the JVM we are running.

I note that go/ej3e says that `CountDownLatch` is great because while mutable, it's not reusable after getting to zero.  `TaskCountDownLatch` specifically violates that.  I'm still correct that `taskRan` isn't thread safe, but we don't get there in this case.  So, we need to look at what JDK/JRE we are using for nightly and see if that's where the problem is.

**Comment by lesv** [comment]
Looking at the configuration and the logs, it's supposed to be the same as Java 8.

**Comment by lesv** [comment]
```
84124ff222e8  latest  2021-05-13T07:34:38  9c4ecd21-da63-4a4b-97c5-9b3e9778a089
gcloud alpha container images list-tags  gcr.io/cloud-devrel-kokoro-resources/java8
```

**Comment by lesv** [comment]
https://cloud.google.com/appengine/docs/standard/java/tools/localunittesting/javadoc/com/google/appengine/tools/development/testing/LocalTaskQueueTestConfig#setdisableautotaskexecution

@averikitsch I hate things that negate a negative.  As I read this, `.setDisableAutoTaskExecution(false)` appears wrong, but
it's possible this is just wrong. :(

**Review by lesv — APPROVED** [review]

**Comment by lesv** [comment]
Yea!

**Pull request merged**